File: tracking.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/tracking.hpp>
#include <iostream>
#include <chrono>

// Variabel global untuk menyimpan koordinat ROI
cv::Rect roi;
bool selecting = false;
bool initialized = false;
cv::Rect roipick;
int fixed_width = 100;
int fixed_height = 100;
cv::Ptr<cv::Tracker> tracker = cv::TrackerCSRT::create();
int frame_count = 0;
double fps = 0.0;
auto start_time = std::chrono::steady_clock::now();

// Callback fungsi mouse untuk memilih ROI
void select_roi(int event, int x, int y, int, void*) {
    if (event == cv::EVENT_LBUTTONDOWN) {
        selecting = true;
    }
    else if (event == cv::EVENT_MOUSEMOVE && selecting) {
        roipick = cv::Rect(x - fixed_width / 2, y - fixed_height / 2, fixed_width, fixed_height);
    }
    else if (event == cv::EVENT_LBUTTONUP) {
        selecting = false;
        roi = cv::Rect(x - fixed_width / 2, y - fixed_height / 2, fixed_width, fixed_height);
    }
}
int main() {
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open camera." << std::endl;
        return -1;
    }
    cap.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 480);
    cv::namedWindow("Frame");
    cv::setMouseCallback("Frame", select_roi);
    cv::Mat frame;
    while (true) {
        cap >> frame;
        if (frame.empty()) {
            break;
        }
        frame_count++;
        auto current_time = std::chrono::steady_clock::now();
        std::chrono::duration<double> elapsed_seconds = current_time - start_time;
        if (elapsed_seconds.count() > 1.0) {
            fps = frame_count / elapsed_seconds.count();
            frame_count = 0;
            start_time = current_time;
        }
        if (selecting) {
            cv::rectangle(frame, roipick, cv::Scalar(255, 0, 0), 2);
        }
        if (!roi.empty() && !initialized) {
            tracker->init(frame, roi);
            initialized = true;
        }
        if (initialized) {
            cv::Rect bbox;
            bool ok = tracker->update(frame, bbox);
            if (ok) {
                cv::rectangle(frame, bbox, cv::Scalar(0, 255, 0), 2, 1);
                int center_x = static_cast<int>(bbox.x + bbox.width / 2);
                int center_y = static_cast<int>(bbox.y + bbox.height / 2);
                std::cout << "FPS " << fps << " Target: (" << center_x << ", " << center_y << ")" << std::endl;
                cv::circle(frame, cv::Point(center_x, center_y), 5, cv::Scalar(0, 0, 255), -1);
                cv::line(frame, cv::Point(center_x, 0), cv::Point(center_x, frame.rows), cv::Scalar(0, 0, 255), 2);  // Vertical line
                cv::line(frame, cv::Point(0, center_y), cv::Point(frame.cols, center_y), cv::Scalar(0, 0, 255), 2);  // Horizontal line
            } else {
                initialized = false;
                roi = cv::Rect();
                std::cout << "FPS " << fps << " Target: Missing" << std::endl;
                cv::putText(frame, "Tracking gagal", cv::Point(100, 80), cv::FONT_HERSHEY_SIMPLEX, 0.75, cv::Scalar(0, 0, 255), 2);
            }
        }
        cv::imshow("Frame", frame);
        int key = cv::waitKey(1) & 0xFF;
        if (key == 27) {  // Tekan ESC untuk keluar
            break;
        } else if (key == 'r') {  // Tekan 'r' untuk reset ROI
            initialized = false;
            roi = cv::Rect();
            selecting = false;
        }
    }
    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: Tahap1/tracking.py
import cv2
import time
roi = None
selecting = False
start_point = (0, 0)
end_point = (0, 0)
cap = cv2.VideoCapture(0)
cv2.namedWindow("Frame")
tracker = cv2.TrackerCSRT.create()
initialized = False
fps = 0
frame_count = 0
start_time = time.time()
def select_roi(event, x, y, flags, param):
    global roi, selecting, start_point, roipick
    w = 100
    h = 100
    if event == cv2.EVENT_LBUTTONDOWN:
        selecting = True
    elif event == cv2.EVENT_MOUSEMOVE and cv2.EVENT_LBUTTONDOWN: 
        roipick = (x - w // 2, y - h // 2, w, h)   
    elif event == cv2.EVENT_LBUTTONUP:
        selecting = False
        roi = (x - w // 2, y - h // 2, w, h)   
while True:
    cv2.setMouseCallback("Frame", select_roi)
    ret, frame = cap.read()
    frame_count += 1
    current_time = time.time()
    elapsed_time = current_time - start_time
    if elapsed_time > 1.0:
        fps = frame_count / elapsed_time
        frame_count = 0
        start_time = current_time
    if not ret:
        break
    if selecting:
        cv2.rectangle(frame, roipick, (255, 0, 0), 2)
    if roi and not initialized:
        tracker.init(frame, roi)
        initialized = True
    if initialized:
        ok, bbox = tracker.update(frame)
        if ok:
            p1 = (int(bbox[0]), int(bbox[1]))
            p2 = (int(bbox[0] + bbox[2]), int(bbox[1] + bbox[3]))
            cv2.rectangle(frame, p1, p2, (0, 255, 0), 2, 1)
            center_x = int(bbox[0] + bbox[2] / 2)
            center_y = int(bbox[1] + bbox[3] / 2)
            print(f"FPS {fps:.2f} Target: ({center_x}, {center_y})")
            cv2.circle(frame, (center_x, center_y), 5, (0, 0, 255), -1)
            cv2.line(frame, (center_x, 0), (center_x, frame.shape[0]), (0, 0, 255), 2)  # Vertical line
            cv2.line(frame, (0, center_y), (frame.shape[1], center_y), (0, 0, 255), 2)  # Horizontal line
        else:
            initialized = False
            roi = None
            print(f"FPS {fps:.2f} Target: Missing")
            cv2.putText(frame, "Tracking gagal", (100, 80), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (0, 0, 255), 2)
    cv2.imshow("Frame", frame)
    key = cv2.waitKey(1) & 0xFF
    if key == 27:  # Tekan ESC untuk keluar
        break
    elif key == ord('r'):  # Tekan 'r' untuk reset ROI
        initialized = False
        roi = None
        selecting = False
cap.release()
cv2.destroyAllWindows()
